pkg/graveler/settings/manager.go

```diff
@@ -0,0 +1,179 @@
+	defaultCacheExpiry     = 1 * time.Second
+)
+
+type cacheKey struct {
+	RepositoryID graveler.RepositoryID
+	Key          string
+}
+
+// Manager is a key-value store for Graveler repository-level settings.
+// Each setting is stored under a key, and can be any proto.Message.
+// Fetched settings are cached using cache.Cache with a default expiry time of 1 second. Hence, the store is eventually consistent.
+type Manager struct {
+	refManager                  graveler.RefManager
+	branchLock                  graveler.BranchLocker
+	blockAdapter                block.Adapter
+	committedBlockStoragePrefix string
+	cache                       cache.Cache
+}
+
+type ManagerOption func(m *Manager)
+
+func WithCacheExpiry(expiry time.Duration) ManagerOption {
+	return func(m *Manager) {
+		m.cache.SetExpiry(expiry)
+		m.cache.SetJitterFn(cache.NewJitterFn(expiry))
+	}
+}
+
+func NewManager(refManager graveler.RefManager, branchLock graveler.BranchLocker, blockAdapter block.Adapter, committedBlockStoragePrefix string, options ...ManagerOption) *Manager {
+	m := &Manager{
+		refManager:                  refManager,
+		branchLock:                  branchLock,
+		blockAdapter:                blockAdapter,
+		committedBlockStoragePrefix: committedBlockStoragePrefix,
+		cache:                       cache.NewCache(cacheSize, defaultCacheExpiry, cache.NewJitterFn(defaultCacheExpiry)),
+	}
+	for _, o := range options {
+		o(m)
+	}
+	return m
+}
+
+// Save persists the given proto.Message under the given repository and key.
+func (m *Manager) Save(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message) error {
+	repo, err := m.refManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	messageBytes, err := proto.Marshal(message)
+	if err != nil {
+		return err
+	}
+	err = m.blockAdapter.Put(ctx, block.ObjectPointer{
+		StorageNamespace: string(repo.StorageNamespace),
+		Identifier:       fmt.Sprintf(settingsSuffixTemplate, m.committedBlockStoragePrefix, key),
+		IdentifierType:   block.IdentifierTypeRelative,
+	}, int64(len(messageBytes)), bytes.NewReader(messageBytes), block.PutOpts{})
+	if err != nil {
+		return err
+	}
+	logger := logging.FromContext(ctx)
+	if logger.IsTracing() {
+		logger.
+			WithField("repo", repositoryID).
+			WithField("key", key).
+			WithField("setting", protojson.Format(message)).
+			Trace("saving repository-level setting")
+	}
+	return nil
+}
+
+// Get fetches the setting under the given repository and key, and returns the result.
+// The result is eventually consistent: it is not guaranteed to be the most up-to-date setting. The cache expiry period is 1 second.
+func (m *Manager) Get(ctx context.Context, repositoryID graveler.RepositoryID, key string, emptyMessage proto.Message) (proto.Message, error) {
+	k := cacheKey{
+		RepositoryID: repositoryID,
+		Key:          key,
+	}
+	message, err := m.cache.GetOrSet(k, func() (v interface{}, err error) {
```

Comment:
Using a struct as a cache key -- this is neat!  I was curious about whether this is guaranteed to work (so no action item requested, just explaining why it works for future reference by future self).  Here's the section in the spec that says it all works: https://golang.org/ref/spec#Map_types. 

>  The comparison operators `==` and `!=` must be fully defined for operands of the key type; thus the key type must not be a function, map, or slice. If the key type is an interface type, these comparison operators must be defined for the dynamic key values; failure will cause a run-time panic. 

Note that this also hints that our `cache.Cache` is not type-safe: it validates that key types are comparable _at runtime_ by panicking.  _This_ usage is, in any case, safe.